File: leetcode_2406/DivideIntervalsIntoMinimumNumberOfGroups.py
class DivideIntervalsIntoMinimumNumberOfGroups:
    def min_number_of_groups(self, intervals: list[list[int]]) -> int:
        start, end = [], []
        for left, right in intervals:
            start.append(left)
            end.append(right)

        start.sort()
        end.sort()

        # Solution 1
        j, group_count = 0, 0
        for st in start:
            if st > end[j]:
                j += 1
            else:
                group_count += 1
        return group_count

        # Solution 2
        # i, j = 0, 0
        # res = 0
        # while i < len(intervals):
        #     if start[i] <= end[j]:
        #         i += 1
        #     else:
        #         j += 1
        #     res = max(res, i-j)
        # return res


def main():
    main_instance = DivideIntervalsIntoMinimumNumberOfGroups()
    print(main_instance.min_number_of_groups([[5, 10], [6, 8], [1, 5], [2, 3], [1, 10]]))


if __name__ == "__main__":
    main()

File: leetcode_2406/DivideIntervalsIntoMinimumNumberOfGroups.java
package leetcode_2406;

import java.util.Arrays;

class DivideIntervalsIntoMinimumNumberOfGroups{
    public int minGroups(int[][] intervals) {
        int n = intervals.length;
        int[] start = new int[n];
        int[] end = new int[n];

        for(int i = 0; i < n; i++)
        {
            start[i] = intervals[i][0];
            end[i] = intervals[i][1];
        }

        Arrays.sort(start);
        Arrays.sort(end);

        // Solution 1
        int j = 0, groupCounts = 0;
        for (int st : start)
        {
            if (st > end[j])
                j++;
            else
                groupCounts++;
        }
        return groupCounts;

        // Solution 2
        // int i = 0, j = 0, res = 0;
        // while(i < n)
        // {
        //     if(start[i] <= end[j])
        //     {   
        //         i++;
        //     }
        //     else
        //     {
        //         j++;
        //     }
        //     res = Math.max(res, i - j);
        // }
        // return res;
    }

    public static void main(String[] args) {
        DivideIntervalsIntoMinimumNumberOfGroups d = new DivideIntervalsIntoMinimumNumberOfGroups();
        System.out.println(d.minGroups(new int[][]{{5, 10}, {6, 8}, {1, 5}, {2, 3}, {1, 10}}));
    }
}